crates/dslab-mp/src/mc/system.rs

```diff
@@ -0,0 +1,67 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use crate::mc::events::McEvent;
+use crate::mc::network::McNetwork;
+use crate::mc::node::{McNode, McNodeState};
+
+pub struct McState {
+    pub node_states: HashMap<String, McNodeState>,
+    pub events: Vec<McEvent>,
+    pub search_depth: u64,
+}
+
+impl McState {
+    pub fn new(events: Vec<McEvent>, search_depth: u64) -> Self {
+        Self {
+            node_states: HashMap::new(),
+            events,
+            search_depth,
+        }
+    }
+}
+
+pub struct McSystem {
+    nodes: HashMap<String, McNode>,
+    net: Rc<RefCell<McNetwork>>,
+    pub(crate) events: Rc<RefCell<Vec<McEvent>>>,
+}
+
+impl McSystem {
+    pub fn new(nodes: HashMap<String, McNode>, net: Rc<RefCell<McNetwork>>, events: Rc<RefCell<Vec<McEvent>>>) -> Self {
+        Self { nodes, net, events }
+    }
+
+    pub fn apply_event(&mut self, event: McEvent) {
+        match event {
+            McEvent::MessageReceived { msg, src, dest } => {
+                self.nodes
+                    .get_mut(self.net.borrow().get_proc_node(&src))
```

Comment:
рантайм падает, потому что не находит процесс на ноде, вроде вот так
```suggestion
                    .get_mut(self.net.borrow().get_proc_node(&dest))
```